test/kad-dht.js

```diff
@@ -1,112 +1,692 @@
+       * |3|       |2|
+       * +-+       +-+
+       */
+      before(function (done) {
+        this.timeout(60 * 1000)
+
+        parallel(nodes.map((node) => (cb) => node.api.id(cb)), (err, ids) => {
+          expect(err).to.not.exist()
+          parallel([
+            (cb) => nodes[3].api.swarm.connect(ids[0].addresses[0], cb),
+            (cb) => nodes[0].api.swarm.connect(ids[1].addresses[0], cb),
+            (cb) => nodes[1].api.swarm.connect(ids[2].addresses[0], cb)
+          ], done)
+        })
+      })
+
+      after(function (done) {
+        this.timeout(80 * 1000)
+        each(nodes, (node, cb) => {
+          node.stop(cb)
+        }, done)
+      })
+
+      it('one hop', function (done) {
+        this.timeout(85 * 1000)
+        const data = crypto.randomBytes(9001)
+
+        addFileAndCat(nodes[0], [nodes[3]], data, done)
+      })
+
+      it('two hops', function (done) {
+        this.timeout(85 * 1000)
+        const data = crypto.randomBytes(9001)
+
+        addFileAndCat(nodes[1], [nodes[3]], data, done)
+      })
+
+      it('three hops', function (done) {
+        this.timeout(85 * 1000)
+        const data = crypto.randomBytes(9001)
+
+        addFileAndCat(nodes[2], [nodes[3]], data, done)
+      })
+    })
+
+    describe('hybrid', () => {
+      let nodes = []
+
+      // Spawn daemons
+      before(function (done) {
+        this.timeout(60 * 1000)
+
+        parallel([
+          (cb) => spawnJsDaemon([], cb),
+          (cb) => spawnGoDaemon([], cb),
+          (cb) => spawnJsDaemon([], cb),
+          (cb) => spawnGoDaemon([], cb)
+        ], (err, spawnedNodes) => {
+          expect(err).to.not.exist()
+          nodes = spawnedNodes
+          done()
+        })
+      })
+
+      // make connections
+      before(function (done) {
+        this.timeout(60 * 1000)
+
+        parallel(nodes.map((node) => (cb) => node.api.id(cb)), (err, ids) => {
+          expect(err).to.not.exist()
+          parallel([
+            (cb) => nodes[3].api.swarm.connect(ids[0].addresses[0], cb),
+            (cb) => nodes[0].api.swarm.connect(ids[1].addresses[0], cb),
+            (cb) => nodes[1].api.swarm.connect(ids[2].addresses[0], cb)
+          ], done)
+        })
+      })
+
+      after(function (done) {
+        this.timeout(80 * 1000)
+        each(nodes, (node, cb) => {
+          node.stop(cb)
+        }, done)
+      })
+
+      it('one hop', function (done) {
+        this.timeout(85 * 1000)
+        const data = crypto.randomBytes(9001)
+
+        addFileAndCat(nodes[0], [nodes[3]], data, done)
+      })
+
+      it('two hops', function (done) {
+        this.timeout(85 * 1000)
+        const data = crypto.randomBytes(9001)
+
+        addFileAndCat(nodes[1], [nodes[3]], data, done)
+      })
+
+      it('three hops', function (done) {
+        this.timeout(85 * 1000)
+        const data = crypto.randomBytes(9001)
+
+        addFileAndCat(nodes[2], [nodes[3]], data, done)
+      })
+    })
+  })
```

Comment:
Again, this batch of tests could be more clean if instead of c&p the tests, they were only written once and then called multiple times with the different node configurations